Revise “Container runtimes” page in light of dockershim removal (#30882)
kubernetes/website: content/en/docs/setup/production-environment/container-runtimes.md

- [CRI-O](#cri-o)
- [Docker](#docker)

{{< note >}}

> Docker container runtime (Dockershim) has been [deprecated](

Reviewer: This implies that Docker container runtime is deprecated and that dockershim and Docker container runtime are the same.
I suggest more clarity that dockershim is deprecated

Reviewer 2: @PranshuSrivastava would you be willing to revise the language here?

Author: @sftim @reylejano sorry I missed this review. I have made the change, PTAL.

> Dockershim, the module that allowed Kubernetes to interact with Docker has been [deprecated](

Reviewer 2: Rather than “has been”, I recommend “was”. The reason is that “has been” suggests it's a recent removal and we prefer the text of documentation to be more timeless (and easier to maintain).

> Dockershim, the module that allowed Kubernetes to interact with Docker has been [deprecated](

Author: @mikebrow  @sftim done.

- [Docker Engine](#docker)
- [Mirantis Container Runtime](#mcr)

{{< note >}}
Dockershim, the portion of code in Kubernetes that provided direct
integration with Docker in prior releases, was removed from Kubernetes
version 1.24. This removal was announced as a [deprecation in Kubernetes v 1.20](
/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
You can check out this [documentation](
/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)

> to understand how this deprecation might affect you. For migrating from

Reviewer 3: Leading "to migrate from dockershim" looks redundant/tautological to me:

For migrating from dockershim ... **to migrate from dockershim.**

Author: done @bart0sh

Reviewer 3: @PranshuSrivastava thanks, looks better now.

> Dockershim, the portion of code in Kubernetes that provided direct
> integration with Docker in prior releases, was removed from Kubernetes
> version 1.24. This removal was announced as a [deprecation in Kubernetes v 1.20](
> /blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
> You can check out this [documentation](
> /docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
> to understand how this deprecation might affect you. To migrate from
> dockershim you can follow [this migration guide](
> /docs/tasks/administer-cluster/migrating-from-dockershim/)
> to migrate from dockershim.

Reviewer 2: ```suggestion
Kubernetes releases before v1.24 included a direct integration with Docker Engine,
using a component named _dockershim_. That special direct integration is no longer
part of Kubernetes (this removal was
[announced](/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
as part of the v1.20 release).

You can read
[Check whether Dockershim deprecation affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/) to understand how this removal might
affect you. To learn about migrating from using dockershim, see
[Migrating from dockershim](/docs/tasks/administer-cluster/migrating-from-dockershim/).
```

Reviewer 4: SGTM

> to migrate from dockershim.

Reviewer 2: If this text is present, the PR must target the _dev-1.24_ branch and not _main_.

/hold
until that discrepancy is dealt with.

Author: @sftim done

Reviewer 2: Thanks
/unhold

however: we may well want a similar-but-different change targeting _main_, so that we can blog about this before the release of v1.24

This PR can still go ahead, I think.

- [Docker Engine](#docker)
- [Mirantis Container Runtime](#mcr)

{{< note >}}

> Docker container runtime (Dockershim) has been [deprecated](

Reviewer: ```suggestion
Docker container runtime (Dockershim) has been [deprecated](
```
Implies that dockershim and Docker container runtime are the same, please revise

> https://kubernetes.io/blog/2020/12/02/dockershim-faq/)

Reviewer 5: ```suggestion
/blog/2020/12/02/dockershim-faq/)
```

as of [version 1.20](

> https://kubernetes.io/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)

Reviewer 5: We use relative paths rather than absolute ones. Also please help make sure the link appears on one line.
```suggestion
/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
```

> and is planned to be removed in by 1.24, therefore to use Docker you need to
> be running an older version of Kubernetes.

Reviewer: ```suggestion
and is planned to be removed in by 1.24, therefore to use Docker you need to
be running an older version of Kubernetes.
```
This is not true, you can use cri-dockerd and use Docker

You can check out this [documentation](

> https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)

Reviewer 5: ```suggestion
/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
```

to understand how this deprecation might affect you. For migrating from

> dockershim you can follow [this](
> https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/)

Reviewer 5: ```suggestion
dockershim you can follow [the migration guide](/docs/tasks/administer-cluster/migrating-from-dockershim/)
```

Reviewer: @PranshuSrivastava check the links from the preview, @tengqm 's suggestion is correct

Author: @reylejano Okay

> document to migrate from dockershim.

Reviewer 5: ```suggestion
to migrate from dockershim.
```

{{< tabs name="tab-cri-containerd-installation" >}}
{{% tab name="Linux" %}}

> 1. Install the `containerd.io` package from the [official containerd website](

Reviewer 6: There are no "containerd.io" packages on the containerd.io website, there it just called "containerd" (and no packages)

The alternative (to "containerd") would be "cri-containerd-cni" which also includes the CRI and CNI configuration:

`/etc/crictl.yaml`

`etc/cni/net.d/10-containerd-net.conflist`

Reviewer 2: Where does the package name `cri-containerd-cni` come from?

Reviewer 3: @sftim cri-containerd-cni is a tarball, recommended as a way to install by the containerd documentation: https://containerd.io/docs/getting-started/#starting-containerd

+1 for not mentioning containerd.io package.

Reviewer 7: makes sense to link to https://containerd.io/downloads/ and https://containerd.io/docs/getting-started/#starting-containerd

Reviewer 3: @afbjorklund installing CNI can cause [hard to debug issues](https://github.com/cri-o/cri-o/issues/2885). I'd recommend to install containerd from `cri-containerd` tarball instead.

Reviewer 6: Kubernetes only works with one CNI at a time, installing podman or cri-o or even containerd is known to interfere with cluster setup...

The recommendation is to wipe /etc/cni before configuring and starting the cluster

Reviewer 3: @afbjorklund  So, what's the point to install CNI if it's going to be removed anyway?

btw, I don't see any recommendation to remove CNI in [this documentation](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/). Can you point me out to the recommendation you've mentioned?

Reviewer 6: I don't think there is any instructions regarding CNI at all ? So it was mostly how I was adding myself.

`rm -f /etc/cni/net.d/*.conf*`

`kubectl apply -f https://raw.githubusercontent.com/flannel-io/flannel/v0.14.0/Documentation/kube-flannel.yml`

This of course also _breaks_ all other usage, so you need to move the podman configuration directory.

```
# Path to the directory where CNI configuration files are located.
#
# network_config_dir = "/etc/cni/net.d/"
```

Similar with the nerdctl configuration, so that it doesn't interfer with Kubernetes' containerd config (of CNI) ?

Reviewer 6: @bart0sh I am not sure what the best way is to package this.

The user _needs_ to set up /etc/crictl.yaml and /etc/cni/net.d

But at the same time, there can be only one container runtime and only one container networking for Kubernetes.

So the files from one will conflict with the files from the other ones. Maybe they should go in /usr/share or in docs ?

Reviewer 3: @afbjorklund installing https://github.com/containerd/containerd/releases/download/v1.6.0/cri-containerd-1.6.0-linux-arm64.tar.gz and then following [this guide](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) is the most straightforward way to get working kubernetes cluster in my opinion.

Reviewer 6: Sure, but it doesn't have any details ? It says you need a CRI and you need a CNI, but after that it is just "GL;HF".

The entire process was more like: https://github.com/lima-vm/lima/blob/v0.8.3/examples/k8s.yaml#L28_L79

Reviewer 3: [this guide](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) was referencing [container runtimes](https://kubernetes.io/docs/setup/production-environment/container-runtimes/) previously IIRC. So, I always followed first guide and it worked for me.

Reviewer 6: Minor packaging detail: runc is bundled with "containerd.io" and with "cri-containerd", there is no "runc.io" package.

There is a regular system "runc" package, that is a dependency of the regular system "containerd" package though.

The docker packages are basically internal to docker, they could change (and have changed before) at any time.

Haven't looked at the rpm packaging (this was deb), but assume there is a similar vendor vs. system duplication ?

Reviewer 2: For SIG Docs: we want to be as generic as we can be about third-party solutions. Even if someone uses containerd, we don't know what OCI runtime they prefer. `runc` is common but there are others that are no less valid.

Feedback on this PR should be aimed at helping the author, @PranshuSrivastava, get this ready for merge. For any other discussion, use issues or Slack.

Also:
```
$ dpkg-query --show runc
runc	1.0.1-0ubuntu2~20.04.1
```
(that's a `.deb`)

@PranshuSrivastava the feedback I'd highlight for you here is: https://github.com/kubernetes/website/pull/30882#discussion_r816258989

>    https://containerd.io/downloads/).Instructions for setting up the Docker

Reviewer 6: You don't need to set up the **Docker** package repository, if installing from the containerd.io.

So either you install the "containerd.io" package from Docker, or you install the "containerd"
package from your distribution (`apt install containerd`) or download it directly from containerd.io

If you want to keep both options (docker.com and containerd.io), they need to be clearly separated.

Reviewer 3: +1

Reviewer 3: @afbjorklund thank you.

Reviewer 7: we should remove the docker repository from the containerd steps and leave the user two options:
- following the guide in https://containerd.io/downloads/ and https://containerd.io/docs/getting-started/#starting-containerd
- use the package manager of the Linux distro

Reviewer 6: Would it be possible to supply deb packages for containerd, the same way as for "cri-tools" and "kubernetes-cni" ?

From apt.kubernetes.io, that is.

Reviewer 7: The package management in k8s needs to be community owned. Currently Google still holds the keys. But also some reorganization of the package repositories might be required for better version flexibility.

The containerd maintainers can open a discussion about that with the k8s SIG Release. In the meantime we should document the best option, IMO.

Reviewer 6: Maybe CRI and CNI should be decoupled from the packaging as well, they would be less outdated that way I think.

Reviewer 2: @PranshuSrivastava I recommend following the approach outlined in https://github.com/kubernetes/website/pull/30882#discussion_r816261724

@afbjorklund, I'm keen to keep the focus of this PR on just the set of changes we'd make to this page so that the v1.24 docs make sense after the dockershim component is removed. Once this merges, we can consider follow-up PRs that make further improvements.

As the proverb doesn't quite say, the perfect is the enemy of the merged.

Reviewer 6: I think changing from docker.com to containerd.io would be enough, go with the tarball if that is all there is.

Reviewer 3: Can we keep containerd installation instructions as they are as they're not actually related to the purpose of this PR. We can update them in a separate PR I believe.

Reviewer 2: After we remove the dockershim, the current instructions are too likely to confuse people. I think fixing that does align with the intent of this PR.

{{< tabs name="tab-cri-containerd-installation" >}}
{{% tab name="Linux" %}}

1. Install the `containerd.io` package from the official Docker repositories.
Instructions for setting up the Docker repository for your respective Linux distribution and
installing the `containerd.io` package can be found at

> [Install Docker Engine](https://docs.docker.com/engine/install/#server).

Author: @reylejano PTAL

> [Install Docker Engine](https://docs.docker.com/engine/install/#server).

Reviewer 4: you can use the containerd release tar provided at containerd.io/downloads.. but that link does not come with an installer or install instructions... more would be needed or linked to in the docs to make that make sense/work (what to do after extracting from the tar for example)

> [Install Docker Engine](https://docs.docker.com/engine/install/#server).

Reviewer 6: It's the same story as cri-dockerd, except that doesn't even have downloads...
(I will pretend that the cri-o instructions are easy to follow and set an example.)

* https://github.com/kubernetes/kubernetes/issues/106917#issuecomment-1023964814

> [Install Docker Engine](https://docs.docker.com/engine/install/#server).

Reviewer 6: I used the normal k8s packages, that installed crictl 1.19 for me

If you want the latest/greatest, you can go for the ones at Kubic

> [Install Docker Engine](https://docs.docker.com/engine/install/#server).

Reviewer 6: @reylejano here are the package dependencies:

Depends: kubelet (>= 1.19.0), kubectl (>= 1.19.0), kubernetes-cni (>= 0.8.7), **cri-tools (>= 1.19.0)**

kubeadm/kubernetes-xenial 1.23.3-00 amd64
kubelet/kubernetes-xenial 1.23.3-00 amd64
kubectl/kubernetes-xenial 1.23.3-00 amd64
kubernetes-cni/kubernetes-xenial 0.8.7-00 amd64
cri-tools/kubernetes-xenial 1.19.0-00 amd64

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl

You still need to configure /etc/crictl.yaml manually

And to make sure to add a CNI plugin (I used flannel)

### Docker Engine {#docker}

> On each of your nodes, install the Docker for your Linux distribution as per

Reviewer: ```suggestion
On each of your nodes, install Docker for your Linux distribution as per
```

Author: done @reylejano

### Docker Engine {#docker}

On each of your nodes, install Docker for your Linux distribution as per
[Install Docker Engine](https://docs.docker.com/engine/install/#server).

> You can find the latest validated version of Docker in this

Reviewer 6: I don't think Kubernetes validates against Docker anymore, and there doesn't seem to be a version in that file.

### Docker Engine {#docker}

> 1. On each of your nodes, install the Docker for your Linux distribution as per

Reviewer 5: ```suggestion
On each of your nodes, install the Docker for your Linux distribution as per
```

> through the supported {{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}}
> (CRI).
> 
> {{< note >}}
> `overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
> or RHEL or CentOS using version 3.10.0-514 and above.
> {{< /note >}}
> 
> - [`cri-dockerd`](https://github.com/Mirantis/cri-dockerd) from Mirantis

Reviewer 2: ```suggestion
You can also find third-party adapters that let you use Docker Engine with Kubernetes
through the supported {{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}}
(CRI).

The following CRI adapters are designed to integrate Docker Engine with CRI clients (such
as Kubernetes v{{< skew currentVersion >}}):

- [`cri-dockerd`](https://github.com/Mirantis/cri-dockerd) from Mirantis

{{< note >}}
If you use Docker Engine on a node that is running Linux kernel version 4.0 or higher, or that
runs RHEL or CentOS and uses the Linux kernel version 3.10.0-514 and above,
the Kubernetes project suggests using the `overlay2` storage driver.
{{< /note >}}
```

Reviewer 4: SGTM

through the supported {{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}}
(CRI).

>    {{< note >}}
>    `overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
>    or RHEL or CentOS using version 3.10.0-514 and above.
>    {{< /note >}}

Reviewer 5: ```suggestion
{{< note >}}
`overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
or RHEL or CentOS using version 3.10.0-514 and above.
{{< /note >}}
```
